app/upgrades/v16/concentrated_pool.go

```diff
@@ -0,0 +1,117 @@
+package v16
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	authkeeper "github.com/cosmos/cosmos-sdk/x/auth/keeper"
+
+	"github.com/osmosis-labs/osmosis/v15/app/keepers"
+	clmodel "github.com/osmosis-labs/osmosis/v15/x/concentrated-liquidity/model"
+	gammkeeper "github.com/osmosis-labs/osmosis/v15/x/gamm/keeper"
+	gammtypes "github.com/osmosis-labs/osmosis/v15/x/gamm/types"
+	"github.com/osmosis-labs/osmosis/v15/x/poolmanager"
+	poolmanagertypes "github.com/osmosis-labs/osmosis/v15/x/poolmanager/types"
+)
+
+// createConcentratedPoolFromCFMM creates a new concentrated liquidity pool with the desiredDenom0 token as the
+// token 0, links it with an existing CFMM pool, and returns the created pool.
+// It uses pool manager module account as the creator of the pool.
+// Returns error if desired denom 0 is not in associated with the CFMM pool.
+// Returns error if CFMM pool does not have exactly 2 denoms.
+// Returns error if pool creation fails.
+func createConcentratedPoolFromCFMM(ctx sdk.Context, cfmmPoolIdToLinkWith uint64, desiredDenom0 string, accountKeeper authkeeper.AccountKeeper, gammKeeper gammkeeper.Keeper, poolmanagerKeeper poolmanager.Keeper) (poolmanagertypes.PoolI, error) {
+	cfmmPool, err := gammKeeper.GetCFMMPool(ctx, cfmmPoolIdToLinkWith)
```

Comment:
We infer denom1 from the balancer liquidity. We provide denom0 because we want to specify which one of the denoms in balancer is desired to be denom0 in CL.

Being able to provide that is desired because of tick spacing